### PR TITLE
Bakerstu cc3220 write active

### PR DESCRIPTION
**Pull request by bakerstu**
- Refactor os_thread_create with weak linked calls for OS-specific threading implementations
- Adds a pthreads based implementation in the cc3220sdk which exposes a pthreads compatible wrapper around freertos threads.
- Cleans up main and sl_Task thread creation to use os_thread_create.
- Removes custom code for handling newlib struct _reent. The earliest freertos that we are supporting (8.2.0)  already has builtin newlib reent functionality.
- Updates select implementation to store the select wakeup index alternatively in the task tag directly or a freertos TLS index. (This is because the task tags are clobbered by the simplelink pthreads implementation.)
- Updates build rules in .mk to use newlib nano.
- updates the logic in CC32xxSocket to manipulate the writeActive and readActive bits only explicitly in select and when an EAGAIN is returned from simplelink. This works around a bug in the simplelink driver that is triggered when sl_Send is called consecutively too frequently.
- ensures that O_NONBLOCK is correctly saved for CC3220 sockets.
- Fixes race conditions in the select wrapper thread.
- Fixes bug when select wakeup between the wlan task and the sl task was not re-tried.

### CODE REVIEW

File: src/freertos_drivers/ti/SimplelinkPthread.cxx [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @file SimpleLinkPthread.cxx
```
Comment (balazsracz) [inline src/freertos_drivers/ti/SimplelinkPthread.cxx]: I don't understand what this file does and why we need it. What does the simplelink SDK break about our implementation of freertos threads? Why can't we have the same implementation that we have been using on all other platforms?
This file seems to have a lot of code copied from os.c in a way that is subtly modified and subject to divergence from each other. There is zero unit testing for all of this code. I have very little confidence that this fork of code is maintainable on the long term and I have a vested interest (just like you) of not having to chase subtle kernel level bugs on the cc3220 platform.

**Comment by balazsracz** [comment]
Please ensure that your PRs have accurate descriptions that relatively exhaustively describe what you are changing. It is very difficult to review a PR where I have to continuously guess why some change is in there and what it is supposed to do.
Also it's not a good idea to mix together debugging changes (like the send/recv hacks) and feature changes (like the newlib reentrant change), because it creates unnecessary coupling of them in the version history: we won't be able to disect or revert one without the other. Please separate those into separate PRs in the future.

**Comment by balazsracz** [comment]
Please test what is the size of a TCB after this change and report your findings.
We had problems before in ensuring that we allocate the appropriate amount of RAM for the struct reent, with the naive solution causing 1K of ram allocated instead of 64 bytes as needed by newlib nano. I do not see the equivalent logic in this PR.

File: src/freertos_drivers/ti/SimplelinkPthread.cxx [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+typedef struct
+{
+    void *(*entry)(void*); ///< thread entry point
+    void *arg; ///< argument to thread
+    TaskList *taskList; //< task list instance
+} OSThreadStartPriv;
+
+/// Entry point to a thread.
+/// @param arg metadata for entering the thread
+static void *os_thread_start(void *arg)
```
Comment (balazsracz) [inline src/freertos_drivers/ti/SimplelinkPthread.cxx]: we decided to platformize this as well with some hooks maybe.

File: include/freertos/FreeRTOSConfig.h [hunk 3]
```diff
@@ -323,6 +324,10 @@ extern unsigned long blinker_pattern;
 #define configSUPPORT_DYNAMIC_ALLOCATION    1
 #endif
 
+/* Enable thread local storage */
+#define configNUM_THREAD_LOCAL_STORAGE_POINTERS 1
+#define TLS_INDEX_SELECT_EVENT_BIT 0
```
[the remark this answers is not in the document]
Comment (bakerstu) [reply include/freertos/FreeRTOSConfig.h]: Done

File: include/freertos/FreeRTOSConfig.h [hunk 4]
```diff
@@ -323,6 +324,10 @@ extern unsigned long blinker_pattern;
 #define configSUPPORT_DYNAMIC_ALLOCATION    1
 #endif
 
+/* Enable thread local storage */
+#define configNUM_THREAD_LOCAL_STORAGE_POINTERS 1
```
[the remark this answers is not in the document]
Comment (bakerstu) [reply include/freertos/FreeRTOSConfig.h]: Done.

[the remark this answers is not in the document]
Comment (bakerstu) [reply include/freertos/FreeRTOSConfig.h]: Task Tag.

File: src/freertos_drivers/common/Select.cxx [hunk 5]
```diff
@@ -45,19 +45,24 @@ static OSEvent wakeup;
 static OSEventType get_event()
 {
     static int thread_count = 0;
-    ThreadPriv *priv = (ThreadPriv*)xTaskGetApplicationTaskTag(NULL);
```
[the remark this answers is not in the document]
Comment (bakerstu) [reply src/freertos_drivers/common/Select.cxx]: Done.

File: src/freertos_drivers/net_cc32xx/CC32xxSocket.cxx [hunk 6]
```diff
@@ -467,10 +480,7 @@ ssize_t CC32xxSocket::recv(int socket, void *buffer, size_t length, int flags)
     {
         s->readActive = false;
     }
-    else
```
[the remark this answers is not in the document]
Comment (bakerstu) [reply src/freertos_drivers/net_cc32xx/CC32xxSocket.cxx]: obsolete.

[the remark this answers is not in the document]
Comment (bakerstu) [reply src/freertos_drivers/net_cc32xx/CC32xxSocket.cxx]: obsolete.

File: src/freertos_drivers/ti/SimplelinkPthread.cxx (hunk 1 above)
Comment (bakerstu) [reply src/freertos_drivers/ti/SimplelinkPthread.cxx]: resolved through discussion.

**Comment by bakerstu** [comment]
I think this is ready for another review.  I applied the output of our discussion and also fixed the _impure_ptr size, verified as 96 bytes.

File: src/freertos_drivers/ti/SimplelinkPthread.cxx [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+    return NULL;
+}
+#else
+#error Unsupported OS
+#endif // __FreeRTOS__
+
+//
+// ::os_thread_create()
+//
+int os_thread_create(os_thread_t *thread, const char *name, int priority,
```
[the remark this answers is not in the document]
Comment (balazsracz) [reply src/freertos_drivers/ti/SimplelinkPthread.cxx]: has this been addressed? Maybe you need to push some commits to github.

File: src/freertos_drivers/ti/SimplelinkPthread.cxx (hunk 2 above)
Comment (bakerstu) [reply src/freertos_drivers/ti/SimplelinkPthread.cxx]: Fixed.

File: src/freertos_drivers/ti/SimplelinkPthread.cxx (hunk 7 above)
[the remark this answers is not in the document]
Comment (bakerstu) [reply src/freertos_drivers/ti/SimplelinkPthread.cxx]: Fixed.

File: src/freertos_drivers/net_cc32xx/CC32xxWiFi.cxx [hunk 8]
```diff
@@ -699,6 +699,21 @@ void CC32xxWiFi::wlan_task()
...
+                --result;
+                portENTER_CRITICAL();
+                SL_SOCKET_FD_CLR(slSockets[i], &wfds);
+                new_highest();
+                CC32xxSocket *s = CC32xxSocket::get_instance_from_sd(slSockets[i]);
+                portEXIT_CRITICAL();
+                /** @todo this is a hack for an sl_Send bandwidth issue in the
+                 * CC3220
+                 */
+                s->writeActiveCnt = 0;
```
[the remark this answers is not in the document]
Comment (bakerstu) [reply src/freertos_drivers/net_cc32xx/CC32xxWiFi.cxx]: code removed.

File: src/freertos_drivers/net_cc32xx/CC32xxSocket.cxx [hunk 9]
```diff
@@ -435,6 +436,18 @@ ssize_t CC32xxSocket::recv(int socket, void *buffer, size_t length, int flags)
         return -1;
     }
 
+    if (s->readActive == false)
+    {
+        errno = EAGAIN;
```
[the remark this answers is not in the document]
Comment (bakerstu) [reply src/freertos_drivers/net_cc32xx/CC32xxSocket.cxx]: code removed.

File: src/freertos_drivers/net_cc32xx/CC32xxSocket.cxx [hunk 10]
```diff
@@ -489,8 +499,33 @@ ssize_t CC32xxSocket::send(int socket, const void *buffer, size_t length, int fl
...
+        {
+            portEXIT_CRITICAL();
+            errno = EAGAIN;
+            return -1;
+        }
+        s->writeActiveCnt = 0;
+        s->writeActive = true;
+    }
+    portEXIT_CRITICAL();
+
```
[the remark this answers is not in the document]
Comment (bakerstu) [reply src/freertos_drivers/net_cc32xx/CC32xxSocket.cxx]: code removed.

File: src/os/os.c [hunk 11]
```diff
@@ -965,21 +935,15 @@ int main(int argc, char *argv[])
...
-                &task_handle);
+                config_main_thread_stack_size() / sizeof(portSTACK_TYPE), NULL,
+                priority, &task_handle);
 #else
-    xTaskGenericCreate(main_thread, (char *)"thread.main",
-                       config_main_thread_stack_size() / sizeof(portSTACK_TYPE), priv,
-                       priority, &task_handle,
-                       (long unsigned int *)stack_malloc(config_main_thread_stack_size()),
-                       NULL);
+#error FREERTOS version v9.0.0 or later required
```
[the remark this answers is not in the document]
Comment (bakerstu) [reply src/os/os.c]: This will be addressed in a pre merge synch with latest master.

File: src/os/os.c [hunk 12]
```diff
@@ -868,40 +886,18 @@ static inline void __attribute__((always_inline)) os_yield_trampoline(void) {
...
-{
-    ThreadPriv *priv = arg;
-    char *argv[2] = {"nmranet", NULL};
-    vTaskSetApplicationTaskTag(NULL, arg);
-    _impure_ptr = priv->reent;
-
-    /* setup the monitoring entries for the timer and idle tasks */
-#if configUSE_TIMERS
-    TaskList *task_new = malloc(sizeof(TaskList) * 2);
-    task_new->task = xTimerGetTimerDaemonTaskHandle();
```
Comment (balazsracz) [inline src/os/os.c]: where did this code go?
I can't quite find it.

File: src/freertos_drivers/net_cc32xx/CC32xxSocket.cxx [hunk 13]
```diff
@@ -492,6 +486,13 @@ ssize_t CC32xxSocket::send(int socket, const void *buffer, size_t length, int fl
         return -1;
     }
 
+    if (s->writeActive == false)
+    {
+        /* typically we would never get here */
+        errno = EAGAIN;
```
Comment (balazsracz) [inline src/freertos_drivers/net_cc32xx/CC32xxSocket.cxx]: I'm adding an assert here that O_NONBLOCK is set.

**Pull request merged**